pkg/controller/azvolumeattachment.go

```diff
@@ -218,6 +235,36 @@ func (r *reconcileAzVolumeAttachment) SyncVolume(ctx context.Context, azVolume v
 		return err
 	}
 
+	// If primary attachment event, reset deletion timestamp
+	if eventType == AttachEvent {
+		if isPrimary {
+			// cancel context of the scheduled deletion goroutine
+			r.cleanUpMapMutex.Lock()
+			cancelFunc, ok := r.cleanUpMap[azVolume.Name]
+			if ok {
+				cancelFunc()
+				delete(r.cleanUpMap, azVolume.Name)
+			}
+			r.cleanUpMapMutex.Unlock()
+		}
+		// If primary detachment event, set deletion timestamp
+	} else if eventType == DetachEvent && azVolume.Spec.MaxMountReplicaCount > 0 {
+		if isPrimary {
+			r.cleanUpMapMutex.Lock()
+			emptyCtx := context.TODO()
+			deletionCtx, cancelFunc := context.WithCancel(emptyCtx)
+			r.cleanUpMap[azVolume.Name] = cancelFunc
+			r.cleanUpMapMutex.Unlock()
+
+			go func(ctx context.Context) {
+				// Sleep
+				time.Sleep(defaultTimeUntilDeletion)
+				_ = r.CleanUpAzVolumeAttachment(ctx, azVolume.Name)
```

Comment:
Wouldn't this cleanup the AzVolumeAttachments after defaultTimeUntilDeletion? Just wondering if there should be a  to avoid the deletion if it is a failover scenario and an attach event comes while waiting.

Reply:
This handled via context cancelling as below.
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/8e8f802c4209d5629e7837c1d7d055a43398065a/pkg/controller/azvolumeattachment.go#L243

Whenever there is a primary attachment deletion, a goroutine will be created that does a scheduled clean up with cancellable context. So if this turns out to be a failover, the context will be cancelled, thereby shutting the clean up goroutine.

Reply:
Makes sense. Thanks for explaining :)